File: src/my_robot_tutorials/scripts/oop_number_counter.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import Int64
from std_srvs.srv import SetBool


class NumberCounter:

	def __init__(self):
		self.counter = 0

		self.number_subscriber = rospy.Subscriber("/number", Int64, self.callback_number)

		self.pub = rospy.Publisher("/number_count", Int64, queue_size=10)

		self.reset_service = rospy.Service("/reset_counter", SetBool, self.callback_reset_counter)

	def callback_number(self, msg):
		self.counter += msg.data
		new_msg = Int64()
		new_msg.data = self.counter
		self.pub.publish(new_msg)

	def callback_reset_counter(self, req):
		if req.data:
			self.counter = 0
			return True, "Counter has been successfully reset"
		return False, "Counter has not been reset"


if __name__ == '__main__':
	rospy.init_node('number_counter')
	NumberCounter()
	rospy.spin()

File: src/my_robot_tutorials/src/oop_number_counter.cpp
#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <std_srvs/SetBool.h>

class NumberCounter {

	private:
	int counter;
	ros::Publisher pub;
	ros::Subscriber number_subscriber;
	ros::ServiceServer reset_service;

	public:
	NumberCounter(ros::NodeHandle *nh) {
		counter = 0;
		
		pub = nh->advertise<std_msgs::Int64>("/number_count", 10);
		
		number_subscriber = nh->subscribe("/number", 1000,
			&NumberCounter::callback_number, this);
		
		reset_service = nh->advertiseService("/reset_counter",
			&NumberCounter::callback_reset_counter, this);
	}

	void callback_number(const std_msgs::Int64& msg) {
		counter += msg.data;
		std_msgs::Int64 new_msg;
		new_msg.data = counter;
		pub.publish(new_msg);
	}

	bool callback_reset_counter(std_srvs::SetBool::Request &req,
								std_srvs::SetBool::Response &res)
	{
		if (req.data) {
			counter = 0;
			res.success = true;
			res.message = "Counter has been successfully reset";
		}
		else {
			res.success = false;
			res.message = "Counter has not been reset";
		}

		return true;
	}
};

int main (int argc, char **argv)
{
	ros::init(argc, argv, "number_counter");
	ros::NodeHandle nh;
	NumberCounter nc = NumberCounter(&nh);
	ros::spin();
}
